src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -154,22 +165,35 @@ doReadRow d k = forModuleNameFix $ \mnFix ->
             then mzero
             -- we merge with (++) which should produce txids most-recent-first
             -- -- we care about the most recent update to this rowkey
-            else MaybeT $ return $! decode $ fromStrict $ DL.head ddata
+            else MaybeT $ return $! decodeStrict' $ DL.head ddata
 
-    lookupInDb :: forall v . FromJSON v => Utf8 -> MaybeT (BlockHandler SQLiteEnv) v
-    lookupInDb rowkey = do
+    lookupInDb
+        :: forall v . FromJSON v
+        => Utf8
+        -> (Utf8 -> BS.ByteString -> MaybeT (BlockHandler SQLiteEnv) v)
+        -> MaybeT (BlockHandler SQLiteEnv) v
+    lookupInDb rowkey checkCache = do
         -- First, check: did we create this table during this block? If so,
         -- there's no point in looking up the key.
         checkDbTableExists tableName
         result <- lift $ callDb "doReadRow"
                        $ \db -> qry db queryStmt [SText rowkey] [RBlob]
         case result of
             [] -> mzero
-            [[SBlob a]] -> MaybeT $ return $! decode $ fromStrict a
+            [[SBlob a]] ->
+              checkCache rowkey a <|>
+              (MaybeT $ return $! decodeStrict' a)
```

Comment:
Should we be reading from sqlite before checking the cache?

Reply:
The cache is not caching the result of the db query, but the result of the result of decoding the row data.

Line 164 reads the row data from the db. Line 184

Reply:
Oh, line 185 is needed for the `noCache`. I guess, it should be moved there.